### PR TITLE
Move email sending for adding a package owner into the service layer

### PR DESCRIPTION
**Pull request by joelverhagen**
Today, all of our emails (except one for AccountDeleter) are send in the ASP.NET Controller layer. This means if another location needs to perform a similar bit of business logic, the email sending code is duplicated.

This PR currently moves a single email send down from the Controller layer into the Service layer. The pieces of the PR of note are:

1. Introduce an `IUrlHelper` abstraction so the service layer can generate URLs.
2. Add `IPackageOwnershipManagementService.AddPackageOwnerWithMessagesAsync`.
3. Use this new method in 2 of the 3 places where owners are added (both of which are already sending mail).

This change also closes a gap in behavior where in one case the newly added owner got an email and in another case they did not. I aligned the behavior so that all owners, new and old, got the notification email.

**Alternatives considered:**

- Adding a new `IPackageOwnershipMessageService` which has this `AddPackageOwnerWithMessagesAsync` method. I thought adding a new method to and existing interface was better.
- Making the `AddPackageOwnerWithMessagesAsync` method have a `string packageUrl` parameter so `IUrlHelper` needn't exist. I didn't like this pattern since some emails that would be sent from the service layer need many URLs and therefore would have many URL parameters on their `*WithMessagesAsync` methods.

Thoughts?

This is related to https://github.com/NuGet/Engineering/issues/3994, which will need these service layer operations to send email.
Progress on https://github.com/NuGet/Engineering/issues/4034.

### CODE REVIEW

File: src/NuGetGallery.Services/PackageManagement/PackageOwnershipManagementService.cs [hunk 1]
```diff
@@ -7,29 +7,54 @@
...
+
+        public async Task AddPackageOwnerWithMessagesAsync(PackageRegistration packageRegistration, User user)
+        {
+            await AddPackageOwnerAsync(packageRegistration, user, commitChanges: true);
+
+            var packageUrl = _urlHelper.Package(packageRegistration.Id, version: null, relativeUrl: false);
+            foreach (var owner in packageRegistration.Owners)
+            {
+                var emailMessage = new PackageOwnerAddedMessage(_appConfiguration, owner, user, packageRegistration, packageUrl);
+                await _messageService.SendMessageAsync(emailMessage);
```
Comment (loic-sharma) [inline src/NuGetGallery.Services/PackageManagement/PackageOwnershipManagementService.cs:56]: Out-of-scope for this PR, but in the future we could make this more performant/reliable by sending all the messages in a single Service Bus batch.

Comment (loic-sharma) [inline src/NuGetGallery.Services/PackageManagement/PackageOwnershipManagementService.cs:56]: I know this is existing code, but do you think it'd be worthwhile to ignore exceptions thrown by `SendMessageAsync` until after all messages have been sent?

Comment (joelverhagen) [reply src/NuGetGallery.Services/PackageManagement/PackageOwnershipManagementService.cs:56]: Agreed. This gets even more gnarly with a bulk admin panel that may send many dozens of messages spanning multiple service layer calls. Almost like we need a MessageContext + `commitChanges: false` like flow to batch together emails. We can consider this later as performance issues arise.

Comment (joelverhagen) [reply src/NuGetGallery.Services/PackageManagement/PackageOwnershipManagementService.cs]: Done.

File: src/NuGetGallery.Services/Providers/IUrlHelper.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Web.Mvc;
+
+namespace NuGetGallery
+{
+    public interface IUrlHelper
+    {
+        string Package(string id, string version, bool relativeUrl);
```
Comment (agr) [inline src/NuGetGallery.Services/Providers/IUrlHelper.cs:15]: Add documentation, please.

**Review by loic-sharma — APPROVED** [review]

File: src/NuGetGallery.Services/Providers/IUrlHelper.cs (hunk 2 above)
Comment (drewgillies) [reply src/NuGetGallery.Services/Providers/IUrlHelper.cs:15]: Yes, I was wondering about this too. I see that this is implemented in code through extension methods. Unsure whether `Package` is a noun or verb here, but looks like the returned string is itself a URI. Anything to clarify this in the interface would be a plus.

Comment (joelverhagen) [reply src/NuGetGallery.Services/Providers/IUrlHelper.cs:15]: I chose to keep the terse method names to mimic the underlying `UrlHelper` extension methods that this class is wrapping.

Totally agree on the docs. Will do.

Comment (joelverhagen) [reply src/NuGetGallery.Services/Providers/IUrlHelper.cs:26]: Done.

**Comment by zhhyu** [comment]
Looking now ...

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 3]
```diff
@@ -2389,26 +2387,6 @@ public virtual ActionResult CancelPendingOwnershipRequest(string id, string requ
...
 
-        /// <summary>
-        /// Send notification that a new package owner was added.
-        /// </summary>
-        /// <param name="package">Package to which owner was added.</param>
-        /// <param name="newOwner">Owner added.</param>
-        private Task SendAddPackageOwnerNotificationAsync(PackageRegistration package, User newOwner)
-        {
-            var packageUrl = Url.Package(package.Id, version: null, relativeUrl: false);
-            Func<User, bool> notNewOwner = o => !o.Username.Equals(newOwner.Username, StringComparison.OrdinalIgnoreCase);
```
Comment (zhhyu) [inline src/NuGetGallery/Controllers/PackagesController.cs:2400]: It's a little weird to see this filter here, and I am not sure whether it's allowed to add a user who is already one of the owners. Do you think whether we need this in the share method?

File: src/NuGetGallery.Services/PackageManagement/PackageOwnershipManagementService.cs [hunk 4]
```diff
@@ -7,29 +7,60 @@
 using System.Collections.Generic;
 using NuGetGallery.Auditing;
 using NuGet.Services.Entities;
+using System.Web.Mvc;
```
Comment (zhhyu) [inline src/NuGetGallery.Services/PackageManagement/PackageOwnershipManagementService.cs:10]: very nit: reorder this library.

File: src/NuGetGallery/Controllers/PackagesController.cs (hunk 3 above)
Comment (joelverhagen) [reply src/NuGetGallery/Controllers/PackagesController.cs:2400]: The purpose of this func is to filter out the owner that was _just_ added as part of this HTTP request. Essentially this code was sending the notification to all pre-existing owners. I opted to remove this in the shared location and instead send the notification to everyone.

File: src/AccountDeleter/Providers/AccountDeleteUrlHelper.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (zhhyu) [inline src/AccountDeleter/Providers/AccountDeleteUrlHelper.cs:8]: Do we need these dependencies here?

File: src/NuGetGallery.Services/Providers/IUrlHelper.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Web.Mvc;
```
Comment (zhhyu) [inline src/NuGetGallery.Services/Providers/IUrlHelper.cs:9]: Do we need these dependencies here?

File: src/NuGetGallery.Services/Providers/IUrlHelper.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (zhhyu) [inline src/NuGetGallery.Services/Providers/IUrlHelper.cs:6]: Do we need these dependencies here?

**Review by zhhyu — APPROVED** [review]

**Pull request merged**